crates/polars-io/src/parquet/write.rs

```diff
@@ -188,8 +188,24 @@ where
 
     /// Write the given DataFrame in the writer `W`. Returns the total size of the file.
     pub fn finish(self, df: &mut DataFrame) -> PolarsResult<u64> {
-        // ensures all chunks are aligned.
-        df.align_chunks();
+        // There are two reasons why we might want to rechunk:
+        //
+        // 1. The different columns have unaligned chunks.
+        // 2. The chunks are aligned, but very small: writing many small chunks
+        //    leads to slow writing performance. We use average estimated
+        //    in-memory size of less than 128KB as a heuristic.
+        //
+        // Neither of these _require_ converting to a single chunk, and
+        // converting to a single chunk has the downside of doubling memory
+        // usage temporarily. In contrast, converting to a series of e.g. 64MB
+        // chunks would give the same benefits without increasing memory usage
+        // meaningfully. However, converting to a single chunk is how this code
+        // already worked previously (covering case 1 only), and implementing
+        // such a chunking algorithm for unaligned chunks is non-trivial, so a
+        // single chunk is what we'll do for now.
+        if df.should_rechunk() || (df.estimated_size() / df.n_chunks() < 128 * 1024) {
+            df.as_single_chunk_par();
```

Comment:
Maybe we can do the rechunking on the `iterator` in line 212. I wonder if that influences the failing test.

Reply:
I'll take a look. Though the fact the test is failing due to `as_single_chunk_par()` being called (something that is possible on `main` too) hints at a bug somewhere.

Reply:
Doing it in the iterator is a problem with just `as_single_chunk_par()`, since you can (and for the example script, do) get lots of single-chunk too-small dataframes. So operating on a single DataFrame isn't sufficient.

I could reuse `StreamingVstacker` from the previous PR, but it would have to be moved into one of the shared crates (`polars-core`? `polars-io`?) and made at least semi-public so it can be used across crates. Do you have a preference?

Reply:
Actually, this raises the question of why `split_df(df, 80)` is resulting in 8500 chunks.

Reply:
```
// we add every chunk as separate dataframe. This make sure that every partition
// deals with it.
```

So maybe there should be a `split_df()` specific to this particular use case.